pwiz_tools/Skyline/Model/TransitionGroup.cs

```diff
@@ -726,14 +726,25 @@ protected override IList<ExplicitLoss> CreateState(ExplicitLoss[] mods)
                 // there should be only one loss
                 TransitionLosses firstLosses = null;
                 List<TransitionLosses> allLosses = null;
-                HashSet<double> allLossMasses = null;
+                HashSet<Tuple<double, int>> allLossMassCharges = null;
                 foreach (var losses in potentialLosses)
                 {
-                    double lossMass = CalcTransitionLossesMass(type, cleavageOffset, massType, losses);
+                    double lossMass = CalcTransitionLossesMass(type, cleavageOffset, massType, losses, out var lossCharge);
+                    // The loss of the charge is only interesting to preserve separate from the mass in
+                    // the case of the precursor, because it has a fixed starting charge. Whereas, fragment
+                    // ions can lose charge during fragmentation no matter what the charge of the loss is,
+                    // and all charges down to charge 1 are allowed for fragments. So, it really is only
+                    // the mass of the loss that matters for fragment ions.
+                    if (type != IonType.precursor)
+                        lossCharge = 0;
                     if (lossMass == 0 ||
-                            (firstLosses != null && firstLosses.Mass == lossMass) ||
-                            (allLossMasses != null && allLossMasses.Contains(lossMass)))
+                        (firstLosses != null && firstLosses.Mass == lossMass &&
+                         firstLosses.TotalCharge == lossCharge) ||

```

Comment:
This "firstLosses.TotalCharge == lossCharge" looks suspicious.

This function is behaving correctly, but I don't think it's doing so for the reasons that you intended.
The only thing that is necessary is that, for fragment ions, if there are multiple losses which have the same mass, this function needs to make sure that the loss with the lowest charge is included in the list that gets returned.

It would make more sense if the logic was:
Does the current loss have the same mass as a loss that we are already planning on returning? If so, compare the losses. If it's a precursor, then keep the current loss if its charge is different than all the ones we already have. If it's not a precursor, then keep the loss with the lowest charge state.

The way the code here is written, there is this other variable "lossCharge" which gets set to zero, but that zero never gets added into the hashSet allLossMassCharges.

Reply:
I see what you mean about the lowest charge. Reversing the order of the losses in the list breaks the test. Okay. Back to work...

Reply:
Okay, added changes to use lowest charge, and added a test the used to fail because of this issue. Hooray for testing!